modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx

```diff
@@ -36,11 +36,23 @@ let counter = 0;
  *
  * @see https://reactjs.org/docs/react-dom.html#render
  */
-export default function render(renderable, renderData, container) {
-	if (!Liferay.SPA || Liferay.SPA.app) {
+export default function render(
+	renderable:
+		| NonNullable<React.ReactNode>
+		| NonNullable<React.ForwardRefExoticComponent<any>>
+		| (() => NonNullable<React.ReactNode>),
+	renderData: {
+		componentId?: string;
+		portletId?: string;
+		[key: string]: unknown;
+	},
+	container: Element | DocumentFragment
+) {
+	if (!(window.Liferay as any).SPA || (window.Liferay as any).SPA.app) {
```

Comment:
So, the goal is to import a `type Liferay` from somewhere even if it's manually generated, or what would be the plan here?

Reply:
What I want to attempt is to define `Liferay` manually in one central location as a so-called "ambient type", which means that it is visible everywhere without a manual import. Makes sense, because globals are visible everywhere without importing them, so the ambient type "matches" that access model.

Ideally, I'll be able to do this in a way that permits augmentation from various places. For example, the central location defines all the necessary types for stuff that is currently scattered across JS and JSP files; then a module like `frontend-js-state-web` that defines `Liferay.State` and is already in TS can augment the type with a simple declaration that looks like `Liferay.State: typeof State`. As we move more and more to TS, there will be less and less manual stuff in the central location. You won't have a single file that you can look at to see "everything that's on `Liferay`", but your IDE will be able to show the type and autocomplete it. We could also generate a `Liferay.d.ts` if we really wanted people to have such a consolidated "manifest", but the truth is we've designed the system in a way that disperses the code all over the place with implicit linkage between the places, so all of this is working to solve a problem of our own creation.

Plan B: If for some reason I can't get the ambient type approach to work (definitely a possibility, because  ambient types are fiddly, and I think they'll be more so with project references thrown into the mix), we can go with the explicit `import type {Liferay} from 'some-explicit-name'`.